x/slashing/keeper/unjail.go

```diff
@@ -30,19 +35,25 @@ func (k Keeper) Unjail(ctx sdk.Context, validatorAddr sdk.ValAddress) error {
 
 	consAddr := sdk.ConsAddress(validator.GetConsPubKey().Address())
 
+	// If the validator has a ValidatorSigningInfo object that signals that the
+	// validator was bonded and so we must check that the validator is not tombstoned
+	// and can be unjailed at the current block.
+	//
+	// A validator that is jailed but has no ValidatorSigningInfo object signals
+	// that the validator was never bonded and must've been jailed due to falling
+	// below their minimum self-delegation. The validator can unjail at any point
+	// assuming they've now bonded above their minimum self-delegation.
```

Comment:
I'd prefer to have some form of context by which we can verify the reason for being jailed for situations like this. In the future it may be that there are other reasons that an unbonded validator may be jailed (and other privileges to being a validator "in waiting"). I could see the Validator `Jailed bool` field being replaced with a `JailedStatus uint8` kind whereby the reason for being jailed could be specified (here, `0` value would probably indicate "not-jailed"). Anyways that's a breaking change, but I think it could be helpful.... probably for another PR or issue!